lib/auth/testauthority/testauthority.go

```diff
@@ -110,12 +111,19 @@ func (n *Keygen) GenerateUserCert(c services.UserCertParams) ([]byte, error) {
 	if !c.PermitPortForwarding {
 		delete(cert.Permissions.Extensions, teleport.CertExtensionPermitPortForwarding)
 	}
-	if len(c.Roles) != 0 {
-		// only add roles to the certificate extensions if the standard format was
-		// requested. we allow the option to omit this to support older versions of
-		// OpenSSH due to a bug in <= OpenSSH 7.1
-		// https://bugzilla.mindrot.org/show_bug.cgi?id=2387
-		if c.CertificateFormat == teleport.CertificateFormatStandard {
+	// Add roles, traits, and route to cluster in the certificate extensions if
+	// the standard format was requested. Certificate extensions are not included
+	// legacy SSH certificates due to a bug in OpenSSH <= OpenSSH 7.1:
+	// https://bugzilla.mindrot.org/show_bug.cgi?id=2387
+	if c.CertificateFormat == teleport.CertificateFormatStandard {
+		traits, err := wrappers.MarshalTraits(&c.Traits)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if len(traits) > 0 {
```

Comment:
Pretty sure this is always `true` since `traits` is going to marshall to `{}` or `null` when empty.

Reply:
What about this? Same for `lib/auth/native.go`.

```go
if len(c.Traits) > 0 {
    traits, err := wrappers.MarshalTraits(&c.Traits)
    if err != nil {
      return nil, trace.Wrap(err)
    }
    cert.Permissions.Extensions[teleport.CertExtensionTeleportTraits] = string(traits)
}

Reply:
the traits were omitted in my tests

Reply:
Just wrote a manual test, and `MarshalTraits` always returned either `{}` or `null` depending on whether the `*wrappers.Traits` value was empty or `nil`... :confused: 

Did your tests verify that the cert extension was actually omitted, or just that no traits were present in the decoded identity?